mlir/lib/Conversion/GPUCommon/GPUOpsLowering.cpp

```diff
@@ -369,7 +370,10 @@ GPUFuncOpLowering::matchAndRewrite(gpu::GPUFuncOp gpuFuncOp, OpAdaptor adaptor,
     }
 
     if (lowersToPointer) {
+      copyPointerAttribute(mlir::NVVM::NVVMDialect::getGridConstantAttrName());
```

Comment:
Having to teach the gpu dialect about target specific dialect attributes isn't ideal, I wonder if this couldn't be served by an interface to allow target-specific dialects to hook into?

Reply:
~I agree, though how many attributes fit the bill of `KernelPointerArgumentAttribute` besides grid constant? Do you think it's still worth adding such an interface?~ If it's a direct violation, then it's probably worth adding 😄 Is `KernelPointerArgumentAttribute` a reasonable name? Any other suggestions?

Reply:
Okay, thanks. 

I wouldn't mind following this pr up with a better fix though--do you have a suggestion for how the proper grid constant attribute would be spelled? I can see how we would attach traits to byval and byref since they are proper attributes, but grid_constant is just a UnitAttr.

Reply:
that's okay 😃 if adding attributes to the nvgpu dialect means I can create a kernel in the GPU dialect that accepts `!llvm.ptr` parameters with the grid constant and byval attributes, then that will work for me! How do you think this should look?

Reply:
I don't see the issue with just landing "llvm pointers, integers, and anything else with a trivial API just gets its argument attributes copied"

Like, yes, in general the ability to ask a dialect if some annotation should survive conversion to LLVM makes sense, but for these purposes, just being able to stick `llvm.byval` or `nvvm.grid_constant` on in higher-level code makes perfect sense.

Reply:
Wait, what's that got to with "just transparently copy the annotations"? What attribute verifier?

(Also, enough GPU compilation flows - ex. Triton - don't use `gpu.func` that actually checking for `gpu.func` is something I'd oppose)

Reply:
> Wait, what's that got to with "just transparently copy the annotations"?

You're copying annotations here from a gpu.func to a llvm.func: these are different ops, an attribute could be designed to be exclusive to gpu.func for example.


> What attribute verifier?

`verifyOperationAttribute ` : https://mlir.llvm.org/docs/DefiningDialects/#discardable-attribute-verification